flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -824,11 +824,14 @@ public ProcessingTimeService getProcessingTimeService() {
 	 * FAILED, and, if the invokable code is running, starts an asynchronous thread
 	 * that aborts that code.
 	 *
-	 * <p>This method never blocks.</p>
+	 * <p>This method never blocks.
 	 */
 	@Override
 	public void handleAsyncException(String message, Throwable exception) {
-		getEnvironment().failExternally(exception);
+		if (isRunning) {
```

Comment:
If `isRunning == true` when entering the if branch, then depending on what happens before `failExternally`, we can assume that the `handleAsyncException` either happened atomically before `isRunning` was set to `false` or not. But what we don't want to happen is that if `isRunning == false`, that we can still fail the task. Thus, I think it solves a valid problem.